quasicrystal/operations.py

```diff
@@ -0,0 +1,155 @@
+import flow
+from flow import environments
+from flow import FlowProject
+import signac
+import numpy as np
+import math
+
+benchmark_name = 'quasicrystal'
+
+potential_k    = 6.25
+potential_phi  = 0.62
+temperature    = 180 * 0.001
+
+# Definition of EOPP potential
+def EOPP(r, rmin, rmax, k, phi):
+    cos = math.cos(k * (r - 1.25) - phi)
+    sin = math.sin(k * (r - 1.25) - phi)
+    V =        pow(r, -15) +       cos * pow(r, -3)
+    F = 15.0 * pow(r, -16) + 3.0 * cos * pow(r, -4) + k * sin * pow(r, -3)
+    return (V, F)
+
+# Determine the cut-off by searching for extrema
+def determineCutoff(k, phi):
+    r = 0.5
+    extremaNum = 0
+    force1 = EOPP(r, 0, 0, k, phi)[1]
+    while (extremaNum < 6 and r < 5.0):
+        r += 0.00001
+        force2 = EOPP(r, 0, 0, k, phi)[1]
+        if (force1 * force2 < 0.0):
+            extremaNum += 1
+        force1 = force2
+    return r
+
+@FlowProject.label
+def pertains(job):
+    global benchmark_name
+    return job.statepoint()['benchmark'] == benchmark_name
+
+@FlowProject.label
+def equilibrated(job):
+    return job.isfile('init.gsd')
+
+# factory method
+def add_equilibration(project):
+    @project.operation('{}-equilibrate'.format(benchmark_name))
+    @project.pre(pertains)
+    @project.post(equilibrated)
+    def equilibrate(job):
+        sp = job.statepoint()
+
+        with job:
+            import hoomd
+            from hoomd import md
+
+            hoomd.context.initialize()
+            rho = 0.03
+            system = hoomd.init.create_lattice(n=sp['n'], unitcell=hoomd.lattice.sc(a=rho**(-1./3.)))
+
+            # generate the pair interaction table
+            cutoff = determineCutoff(potential_k, potential_phi)
+            nl = md.nlist.cell()
+            table = md.pair.table(nlist = nl, width = 1000)
+            table.pair_coeff.set('A', 'A', func = EOPP, rmin = 0.5, rmax = cutoff,
+                                 coeff = dict(k = potential_k, phi = potential_phi))
+
+            # Integrate at constant temperature
+            md.integrate.nvt(group = hoomd.group.all(), tau = 1.0, kT = temperature)
+            md.integrate.mode_standard(dt = 0.01)
+
+            hoomd.run(100000)
+            hoomd.dump.gsd('init.gsd', period=None, overwrite=True,group=hoomd.group.all())
+
+def add_benchmark(project, mode, nranks, gpu_ids=[]):
+    name = ''
```

Comment:
This name logic is also repeated quite a bit. A utility function might be worth it.